### PR TITLE
Added Shims for adding Databricks 14.3 Support [databricks]

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds Shims for Databricks 14.3

### CODE REVIEW

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
@gerashegalov @mythrocks PTAL

File: sql-plugin/src/main/spark350db/scala/com/nvidia/spark/rapids/spark350db/RapidsShuffleManager.scala
Comment (gerashegalov) [inline sql-plugin/src/main/spark350db/scala/com/nvidia/spark/rapids/spark350db/RapidsShuffleManager.scala:1]: This file should be dropped after upmerging to branch-24.12

File: delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransactionBase.scala
Comment (gerashegalov) [inline delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransactionBase.scala:1]: Luckily git dedupes blobs but I wish we would not have 3 copies of these files in our repo

**Review by gerashegalov — CHANGES_REQUESTED** [review]
need to upmerge

File: datagen/src/main/spark320/scala/org/apache/spark/sql/tests/datagen/DataGenExprShims.scala [hunk 1]
```diff
@@ -36,6 +36,7 @@
 {"spark": "342"}
 {"spark": "343"}
 {"spark": "350"}
+{"spark": "350db"}
```
[the remark this answers is not in the document]
Comment (razajafri) [reply datagen/src/main/spark320/scala/org/apache/spark/sql/tests/datagen/DataGenExprShims.scala:39]: Good foresight. Will make the change

File: delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransactionBase.scala
Comment (razajafri) [reply delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransactionBase.scala:1]: Do we want to make the decision to shim these? @jlowe I know there is a reason you wanted to keep it the way it is today. Any reason we can't/shouldn't use shims?

Comment (jlowe) [reply delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransactionBase.scala:1]: We would want to shim these if possible to reduce the duplication.

Comment (razajafri) [reply delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuOptimisticTransactionBase.scala:1]: I would like to do that as a part of a follow-on if that's ok

**Comment by razajafri** [comment]
@gerashegalov I have addressed your concerns. PTAL

File: sql-plugin/src/main/spark350db143/scala/com/nvidia/spark/rapids/shims/spark350db143/SparkShimServiceProvider.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+{"spark": "350db143"}
+spark-rapids-shim-json-lines ***/
+package com.nvidia.spark.rapids.shims.spark350db143
+
+import com.nvidia.spark.rapids.{DatabricksShimVersion, ShimVersion}
+
+import org.apache.spark.SparkEnv
+
+object SparkShimServiceProvider {
+  val VERSION = DatabricksShimVersion(3, 5, 0)
```
Comment (gerashegalov) [inline sql-plugin/src/main/spark350db143/scala/com/nvidia/spark/rapids/shims/spark350db143/SparkShimServiceProvider.scala:27]:
```suggestion
  val VERSION = DatabricksShimVersion(3, 5, 0, "14.3")
```

File: sql-plugin/src/main/spark350db143/scala/com/nvidia/spark/rapids/shims/spark350db143/SparkShimServiceProvider.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+class SparkShimServiceProvider extends com.nvidia.spark.rapids.SparkShimServiceProvider {
+
+  override def getShimVersion: ShimVersion = SparkShimServiceProvider.VERSION
+
+  def matchesVersion(version: String): Boolean = {
+    SparkEnv.get.conf.get("spark.databricks.clusterUsageTags.sparkVersion", "").startsWith("14.3.")
```
Comment (gerashegalov) [inline sql-plugin/src/main/spark350db143/scala/com/nvidia/spark/rapids/shims/spark350db143/SparkShimServiceProvider.scala:35]:
```suggestion
    DatabricksShimServiceProvider.matchesVersion("14.3.x")
```

File: sql-plugin/src/main/spark320/scala/com/nvidia/spark/rapids/RapidsShuffleManager.scala
Comment (gerashegalov) [inline sql-plugin/src/main/spark320/scala/com/nvidia/spark/rapids/RapidsShuffleManager.scala:1]: Update copyright

**Review by gerashegalov — COMMENTED** [review]
LGTM. at least one copyright to update

**Review by gerashegalov — DISMISSED** [review]

**Comment by gerashegalov** [comment]
Re failing [package-tests (350db143, false)](https://github.com/NVIDIA/spark-rapids/actions/runs/11671607945/job/32498487969?pr=11635#logs) ,  need to make sure that the db buildvers are not included in the Github Workflow build matrix

**Comment by razajafri** [comment]
> Re failing [package-tests (350db143, false)](https://github.com/NVIDIA/spark-rapids/actions/runs/11671607945/job/32498487969?pr=11635#logs) , need to make sure that the db buildvers are not included in the Github Workflow build matrix

Hang on, I will have to modify the workflow file and remove the change from pom.xml

**Comment by gerashegalov** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM

**Comment by razajafri** [comment]
build

**Review by mythrocks — APPROVED** [review]

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
> @razajafri Can you enable Databricks-14.3 in the pre-merge CI pipeline in this PR as below?
> 
> https://github.com/NVIDIA/spark-rapids/blob/branch-24.12/jenkins/Jenkinsfile-blossom.premerge-databricks#L94
> 
> `values '11.3', '12.2', '13.3', '14.3'`

I have updated it but this will turn the integration tests on. We don't want the integration tests to run against Databricks 14.3 because we aren't ready.

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
build

**Comment by razajafri** [comment]
> build

AWS_INSUFFICIENT_INSTANCE_CAPACITY_FAILURE

**Comment by razajafri** [comment]
build

**Comment by pxLi** [comment]
build

**Comment by pxLi** [comment]
> > build
> 
> AWS_INSUFFICIENT_INSTANCE_CAPACITY_FAILURE

this change requires 8 clusters (4 shims x 2 parallelism) for the CI, we are considering using other instance types or temply disable parallel run for databricks runtimes.

**Review by gerashegalov — APPROVED** [review]

**Pull request merged**